keygen/src/main.c

```diff
@@ -0,0 +1,231 @@
+
+    /* Prompt the user for a password */
+    get_pass(&login.password);
+
+    /* Only continue if all required parameters have been given */
+    if (!ipv6_network || !enftun_cfg_file || !login.username || !login.password) {
+        print_help(argv[0]);
+        goto out;
+    }
+
+    /* Get the enftun config */
+    enftun_config_init(cfg);
+    enftun_config_parse(cfg, enftun_cfg_file);
+
+    if (!prompt_user(login.username, ipv6_network, enftun_cfg_file, cfg->cert_file, cfg->key_file)) {
+        printf("Aborting...\n");
+        goto out;
+    }
+
+    /* Log in with XCR */
+    printf("Logging in... ");
+    ret = xcr_send_login(&login, &xcr_auth);
+    if (!ret) {
+        printf("Failed. Exiting.\n");
+        goto out;
+    }
+    printf("Successful.\n");
+
+    /* Seed srand for the encryption functions */
+    srand(time(NULL));
+
+    /* Generate a NID_X9_62_prime256v1 cert */
+    printf("Generating EC Key... ");
+    pkey = gen_EC_PKEY();
+    if (!pkey) {
+        printf("Failed. Exiting.\n");
+        goto out;
+    }
+    printf("Successful.\n");
+
+    /* Create a request to have IAM create a new endpoint */
+    printf("Preparing IAM request... ");
+    ep_req = iam_new_ep_auth_network(ipv6_network, public_key_encode_x962(pkey));
+    if (!ep_req) {
+        printf("Failed. Exiting.\n");
+        goto out;
+    }
+    printf("Successful.\n");
+
+    /* Send the request to IAM */
+    printf("Requesting IP address... ");
+    ret = iam_send_ep_auth(ep_req, xcr_auth.token, &ep_resp);
+    if (!ret) {
+        printf("Failed. Exiting.\n");
+        goto out;
+    }
+    printf("Successful. (%s)\n", ep_resp.address);
+
+    /* Generate the certificate with the returned address */
+    printf("Creating signed cert... ");
+    cert = gen_signed_cert(pkey, ep_resp.address);
+    if (!cert) {
+        printf("Failed. Exiting.\n");
+
+        /* TODO Unprovision */
+        printf("Warning: IP Address %s is still provisioned.\n", ep_resp.address);
+        goto out;
+    }
+    printf("Successful. (%s)\n", ep_resp.address);
+
+    /* Write the cert and key for enftun */
+    printf("Saving signed cert... ");
+    ret = write_x509_cert(cert, cfg->cert_file);
+    if (!ret) {
+        printf("Failed. Exiting.\n");
+        /* TODO Unprovision */
+        printf("Warning: IP Address %s is still provisioned.\n", ep_resp.address);
+        goto out;
+    }
+    printf("Cert saved to %s\n", cfg->cert_file);
+
+    printf("Saving private key... ");
+    ret = write_EC_private_key(pkey, cfg->key_file);
+    if (!ret) {
+        printf("Failed. Exiting.\n");
+        /* TODO Unprovision */
+        printf("Warning: IP Address %s is still provisioned.\n", ep_resp.address);
+        goto out;
+    }
+    printf("Key saved to %s\n", cfg->key_file);
+
+    printf("All operations were successful.\n");
+
+    /* Clean up */
+    destroy_EC_PKEY(pkey);
+    ep_auth_resp_destroy(&ep_resp);
+
```

Comment:
`cert` needs to be `free`'d, too

Reply:
Made irrelevant due to other changes. Cert doesn't leave `write_cert` now.